[Workers] Runtime API: TCP Socket API Docs `connect()` (#8795)
cloudflare/cloudflare-docs: content/workers/runtime-apis/connect.md

---
pcx_content_type: configuration

> title: Connect (TCP Socket API)

Author: This ends up being a bit long in the sidebar nav, but I think important since people otherwise don't know the `connect()` API by name yet.

Reviewer: I would argue that for this section the title should not even have the word `connect`.
There are a number of other pages in this section where the title is more descriptive - e.g. "Durable Objects", "Service bindings", "Node compatibility".
So instead, I think this title could just be "TCP sockets" or similar. Then in the page itself introduce the actual function call that gets you a socket (i.e. `connect()`).

  - `off` — do not use TLS.
  - `on` — use TLS.
  - `starttls` — do not use TLS initially, but allow the socket to be upgraded to use TLS by calling [`startTls()`](/workers/runtime-apis/connect/#how-to-implement-the-starttls-pattern).

- `allowHalfOpen` {{<type>}}boolean{{</type>}} — Defaults to `false`
  - Defines whether the writable side of the TCP socket will automatically close on EOF. When set to `false`, the writable side of the TCP socket will automatically close on EOF. When set to `true`, the writable side of the TCP socket will remain open on EOF.
  - This option allows for interoperability with TCP socket implementations in Node.js and Deno, which currently behave differently.

{{</definitions>}}

### `Socket`

{{<definitions>}}

- {{<code>}}readable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/readablestream/">}}ReadableStream{{</type-link>}}
  - Returns the readable side of the TCP socket.

- {{<code>}}writeable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/writeable/">}}WriteableStream{{</type-link>}}
  - Returns the writable side of the TCP socket.

- `closed()` {{<type>}}`Promise<void>`{{</type>}}

>   - TODO

Author: ```suggestion
  - Used to check if the socket has been closed
```

@dom96 — is this true? (I actually wasn't immediately sure offhand how one might use this, imagine you have a clearer example than I do)

Reviewer: Rather than "check" if the socket is closed, I would say this is a way to run some clean up "when" the socket is closed. The promise will only be fulfilled (e.g. resolved) once the socket closes. It is not something you can poll.

- `close()` {{<type>}}`Promise<void>`{{</type>}}
  - Closes the TCP socket. If [`allowHalfOpen`](/workers/runtime-apis/connect/#socketoptions) is set to `true`, only the writeable side of the socket will be closed, and the readable side of the socket will remain readable.

- {{<code>}}startTls(options?: {{<prop-meta>}}optional{{</prop-meta>}} {{<type-link href="/workers/runtime-apis/connect/#tlsoptions">}}TlsOptions{{</type-link>}}){{</code>}} : {{<type-link href="/workers/runtime-apis/connect/#socket">}}Socket{{</type-link>}}
  - Upgrades an insecure socket to a secure one that uses TLS, returning a new [Socket](/workers/runtime-apis/connect#socket). Note that in order to call `startTls()`, you must set [`secureTransport`](/workers/runtime-apis/connect/#socketoptions) to `starttls` when initially calling `connect()` to create the socket.

{{</definitions>}}

### `TlsOptions`

{{<definitions>}}

- `expectedServerHostname` {{<type>}}string{{</type>}}

>   - TODO

Author: @dom96 — this is about [SNI](https://www.cloudflare.com/learning/ssl/what-is-sni/)? Can you drop in a brief explainer and I can edit?

https://github.com/cloudflare/workerd/blob/main/src/workerd/api/sockets.c%2B%2B#L210

Author: https://github.com/cloudflare/cloudflare-docs/pull/8795/commits/40a6dd10d3ba69787d78ac6f4a5868899c738b51

- `secureTransport` {{<type>}}"off" | "on" | "starttls"{{</type>}} — Defaults to `off`
  - Specifies whether or not to use [TLS](https://www.cloudflare.com/learning/ssl/transport-layer-security-tls/) when creating the TCP socket.
  - `off` — do not use TLS.
  - `on` — use TLS.
  - `starttls` — do not use TLS initially, but allow the socket to be upgraded to use TLS by calling [`startTls()`](/workers/runtime-apis/connect/#how-to-implement-the-starttls-pattern).

- `allowHalfOpen` {{<type>}}boolean{{</type>}} — Defaults to `false`
  - Defines whether the writable side of the TCP socket will automatically close on EOF. When set to `false`, the writable side of the TCP socket will automatically close on EOF. When set to `true`, the writable side of the TCP socket will remain open on EOF.
  - This option is similar to that offered by the Node.js [`net` module](https://nodejs.org/api/net.html) and allows interoperability with code which utilises it.

{{</definitions>}}

### `Socket`

{{<definitions>}}

- {{<code>}}readable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/readablestream/">}}ReadableStream{{</type-link>}}
  - Returns the readable side of the TCP socket.

> - {{<code>}}writeable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/writeable/">}}WriteableStream{{</type-link>}}

Author: ```suggestion
- {{<code>}}writeable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/writablestream/">}}WriteableStream{{</type-link>}}
```

---
pcx_content_type: configuration
title: Connect (TCP Socket API)
weight: 4
---

# connect()

> The `connect()` API provides a programming interface for creating an outbound [TCP connection](https://www.cloudflare.com/learning/ddos/glossary/tcp-ip/) from a Cloudflare Worker. It returns a TCP socket, with both a [readable](/workers/runtime-apis/streams/readablestream/) and [writable](/workers/runtime-apis/streams/writablestream/) stream of data, allowing you to read and write data on an ongoing basis, as long as the connection remains open.

Author: ```suggestion
The Workers runtime provides an API, `connect()`, for creating a outbound [TCP connections](https://www.cloudflare.com/learning/ddos/glossary/tcp-ip/) from Workers. It returns a TCP socket, with both a [readable](/workers/runtime-apis/streams/readablestream/) and [writable](/workers/runtime-apis/streams/writablestream/) stream of data, allowing you to read and write data on an ongoing basis, as long as the connection remains open.
```

`connect()` is provided as a [Runtime API](/workers/runtime-apis/), and is accessed by importing the `connect` function from `cloudflare:sockets`, akin to how one imports built-in modules in Node.js. A simple example of creating a TCP socket, writing to it, and returning the readable side of the socket as a response, is shown below:

```typescript
import { connect } from 'cloudflare:sockets';

const connectionUrl = "<URL>";

export default {
  async fetch(req, env) {

    try {
      const socket = connect(connectionUrl);

      const writer = socket.writable.getWriter()
      const encoder = new TextEncoder();
      const encoded = encoder.encode(url.pathname + "\r\n");
      await writer.write(encoded);

      return new Response(socket.readable, { headers: { "Content-Type": "text/plain" } });
    } catch (error) {
      return new Response("Socket connection failed: " + error, { status: 500 });
    }
  }

> };

Author: Updated

  - `on` — use TLS.
  - `starttls` — do not use TLS initially, but allow the socket to be upgraded to use TLS by calling [`startTls()`](/workers/runtime-apis/connect/#how-to-implement-the-starttls-pattern).

- `allowHalfOpen` {{<type>}}boolean{{</type>}} — Defaults to `false`
  - Defines whether the writable side of the TCP socket will automatically close on EOF. When set to `false`, the writable side of the TCP socket will automatically close on EOF. When set to `true`, the writable side of the TCP socket will remain open on EOF.
  - This option allows for interoperability with TCP socket implementations in Node.js and Deno, which currently behave differently.

{{</definitions>}}

### `Socket`

{{<definitions>}}

- {{<code>}}readable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/readablestream/">}}ReadableStream{{</type-link>}}
  - Returns the readable side of the TCP socket.

- {{<code>}}writeable(){{</code>}} : {{<type-link href="/workers/runtime-apis/streams/writeable/">}}WriteableStream{{</type-link>}}
  - Returns the writable side of the TCP socket.

- `closed()` {{<type>}}`Promise<void>`{{</type>}}
  - TODO

- `close()` {{<type>}}`Promise<void>`{{</type>}}

>   - Closes the TCP socket. If [`allowHalfOpen`](/workers/runtime-apis/connect/#socketoptions) is set to `true`, only the writeable side of the socket will be closed, and the readable side of the socket will remain readable.

Author: Thinking back to the Node.js example you'd shared awhile back:

```javascript
socket.end(); // Same as writer.close() in Deno.
 
// The socket's writable side has been closed,  so now
// it is in a half-closed state. The data sent by the
// server will still be readable.
```

I gathered from this that the `allowHalfOpen` option exists to mirror this behavior? Where one can call `close()` or `end()` and still have the socket remain "half open", with the readable side still able to be read?

If not, then why does `allowHalfOpen` exist?

>   - Closes the TCP socket. If [`allowHalfOpen`](/workers/runtime-apis/connect/#socketoptions) is set to `true`, only the writeable side of the socket will be closed, and the readable side of the socket will remain readable.

Author: Ah I see: https://github.com/cloudflare/cloudflare-docs/pull/8795/files#r1188560682

- {{<code>}}startTls(options?: {{<prop-meta>}}optional{{</prop-meta>}} {{<type-link href="/workers/runtime-apis/connect/#tlsoptions">}}TlsOptions{{</type-link>}}){{</code>}} : {{<type-link href="/workers/runtime-apis/connect/#socket">}}Socket{{</type-link>}}
  - Upgrades an insecure socket to a secure one that uses TLS, returning a new [Socket](/workers/runtime-apis/connect#socket). Note that in order to call `startTls()`, you must set [`secureTransport`](/workers/runtime-apis/connect/#socketoptions) to `starttls` when initially calling `connect()` to create the socket.

{{</definitions>}}

### `TlsOptions`

{{<definitions>}}

- `expectedServerHostname` {{<type>}}string{{</type>}}
  - TODO

{{</definitions>}}

## How to implement the startTLS pattern

> Many TCP-based systems, including databases and email servers, require that clients use a pattern called StartTLS when connecting. In this pattern, the client first creates an insecure TCP socket, without TLS, and then "upgrades" it to a secure TCP socket, that uses TLS. The `connect()` API simplifies this by providing a method, `startTls()`, which returns a new `Socket` instance that uses TLS:

Author: [https://github.com/cloudflare/cloudflare-docs/pull/8795/commits/b4086f8172fe5ecce8ade5e0ac8733f031c9816fI hear your feedback on this. Also want it to be discoverable based on name. Most API docs are n](https://github.com/cloudflare/cloudflare-docs/pull/8795/commits/b4086f8172fe5ecce8ade5e0ac8733f031c9816f)

> ```typescript

Author: Sorting out details on that

> ```typescript

Author: Would be nice but don't think there's anything usable here — would get hammered with connections.

```typescript
import { connect } from "cloudflare:sockets"

const socket = connect("my-url.com:70");
const reader = socket.readable.getReader();
socket.close();

// After close() is called, you can no longer read from the readable side of the socket
const reader = socket.readable.getReader(); // This fails
```

If you need to create a socket that is compatible with existing code that assumes the pattern from Node.js, where the readable side of a socket remains open after `socket.end()` is called, you can set the `allowHalfOpen` option when creating a socket.

```typescript
import { connect } from "cloudflare:sockets"

const socket = connect("my-url.com:70", { allowHalfOpen: true });
const reader = socket.readable.getReader();
socket.close();

// After close() is called, you can continue to read from the readable side of the socket
const reader = socket.readable.getReader(); // This works
```

Note the following:

- `startTls()` can only be called if `secureTransport` is set to `starttls` when creating the initial TCP socket.

> - Once `startTls()` is called, the initial socket is closed and no longer usable. In the example above, one would would use the newly created `secureSocket`.

Reviewer 2: This feels a bit out of place, we already explain this above as well, maybe we can just remove this?

const secureSocket = socket.startTls();
```

## Error handling

> - TODO - show what happens when the server closes the connection, and how to handle it gracefully.

Author: @dom96 — can you drop in an example?
